File: mountain/cloudrip-siege.lua
-- level 2
-- using "bash", "move"
function pos2xy(pos) return pos['x'], pos['y'] end

tp = 1
towerPos = {{x=68, y=41}, {x=50, y=41}, {x=77, y=41}, {x=32, y=41}, {x=23, y=41}}

function dist2(a, b)
    local x, y = a.x - b.x, a.y - b.y
    return x*x + y*y
end
function bestCoin(xs)
    local r, maxR = nil, 0
    for i = 1, #xs do
        local v = xs[i].value/self:distanceTo(xs[i])
        if v > maxR then
            r, maxR = xs[i], v
        end
    end
    return r
end

loop
    local c = bestCoin(self:findItems())
    local e = self:findNearest(self:findEnemies())
    if e then
        if self:distanceTo(e) <= 10 and e.maxHealth > 200 and self:isReady("bash") then
            self:bash(e)
        end
    end
    if self.gold >= self:costOf("arrow-tower") then
        if dist2(self.pos, towerPos[tp]) > 100 then
            self:move(towerPos[tp])
        else
            self:buildXY("arrow-tower", pos2xy(towerPos[tp]))
            if tp == #towerPos then
                tp = 1
            else
                tp = tp + 1
            end
        end
    elseif c then
        self:move(c.pos)
    end
end

File: mountain/cloudrip-siege.clj
(def t 0)
(def towers [[68 41][50 41][77 41][32 41][23 41]])

(defn dist2 [xs ys]
  (def x (- (.x xs) (first ys)))
  (def y (- (.y xs) (second ys)))
  (+ (* x x) (* y y)))

(while true
  (def e (.findNearest this (.findEnemies this)))
  (if e
    (if (and (< (.distanceTo this e) 10) (.isReady this "bash"))
        (.bash this e)))
  (if (>= (.gold this) (.costOf this "arrow-tower"))
    (when true
      (def nt (nth towers t))
      (if (> (dist2 (.pos this) nt) 100)
        (.move this (clj->js { :x (first nt), :y (second nt) }))
        (when true
          (.buildXY this "arrow-tower" (first nt) (second nt))
          (def t (mod (inc t) (count towers))))))
    (when true
      (def is (.findItems this))
      (def c nil)
      (when (seq is)
        (def cv 0)
        (def i 0)
        (while (< i (count is))
          (if (> (/ (.value (nth is i)) (.distanceTo this (nth is i))) cv)
            (when true
              (def c (nth is i))
              (def cv (/ (.value (nth is i)) (.distanceTo this (nth is i))))))
          (def i (inc i))))
      (if c
        (.move this (.pos c))
        (.wait this 0.01)))))
